XenModel/Actions/VMAppliances/AssignVMsToVMApplianceAction.cs

```diff
@@ -57,11 +69,14 @@ protected override void Run()
             Description = Messages.ASSIGNING_VM_APPLIANCE;
             foreach (var xenRef in _vmAppliance.VMs)
             {
-                VM.set_appliance(Session, xenRef, null);
+                var found = _selectedVMs.Any(vm => vm.opaque_ref == xenRef);

```

Comment:
This is an O(n^2) operation. I think it must be possible to take the set difference between _selectedVMs and _vmAppliance.VMs more efficiently.